### PR TITLE
(PUP-9076) Do not push nil on server context

### PR DESCRIPTION
**Pull request by melissa**

### CODE REVIEW

**Comment by melissa** [comment]
@joshcooper after my lack of success repro'ing https://tickets.puppetlabs.com/browse/PUP-9076, I think the scope of this ticket should just be to smooth over some of the rough edges with `server` vs. `server_list`. After reading through our grooming notes, and thinking about it for a bit, these changes encompass a lot of what we had been talking about, I think. I still need to do a lot more testing with this, but I wanted to get a feel for if I'm pivoting correctly on this or not

**Comment by melissa** [comment]
jenkins please test this

**Comment by joshcooper** [comment]
I can repro in 5.5.x using:

```
$ bundle exec puppet agent -td --server_list neptune --server mars
Debug: Creating new connection for https://neptune:8140
Debug: Using cached certificate_revocation_list for ca
Debug: Starting connection for https://neptune:8140
Warning: Could not select a functional puppet master
Debug: node supports formats: json msgpack pson yaml
Debug: Creating new connection for https://:0
Debug: Starting connection for https://:0
Warning: Unable to fetch my node definition, but the agent run will continue:
Warning: Failed to open TCP connection to :0 (Can't assign requested address - connect(2) for nil port 0)
...
```
Using your PR, it correctly aborts the run if all servers in the list have been exhausted.

```
$ bundle exec puppet agent -td --server_list neptune --server mars
...
Debug: Creating new connection for https://neptune:8140
Debug: Using cached certificate_revocation_list for ca
Debug: Starting connection for https://neptune:8140
Error: Could not run Puppet configuration client: Could not select a functional puppet master from server_list
```

One suggestion is to include the list servers we tried to connect to (including their port numbers) in the error message above.

File: lib/puppet/configurer.rb [hunk 1]
```diff
@@ -222,8 +222,7 @@ def run(options = {})
           found = find_functional_server()
           server = found[:server]
           if server.nil?
-            Puppet.warning _("Could not select a functional puppet master")
-            server = [nil, nil]
+            raise Puppet::Error, _("Could not select a functional puppet master from server_list")
```
Comment (joshcooper) [inline lib/puppet/configurer.rb]: How about appending `Puppet.settings[:server_list]` to the error message, so we know what servers it tried?

**Comment by melissa** [comment]
```
$  be ./bin/puppet agent -t --server_list 'neptune,pluto,uranus' --server mars
Error: Could not run Puppet configuration client: Could not select a functional puppet master from server_list: '[["neptune", 8140], ["pluto", 8140], ["uranus", 8140]]'
```

File: lib/puppet/configurer.rb (hunk 1 above)
Comment (melissa) [reply lib/puppet/configurer.rb]: Done

File: spec/unit/configurer_spec.rb [hunk 2]
```diff
@@ -1027,17 +1027,15 @@ def expects_neither_new_or_cached_catalog
...
       pool = Puppet::Network::HTTP::Pool.new(Puppet[:http_keepalive_timeout])
       Puppet::Network::HTTP::Pool.expects(:new).returns(pool)
       Puppet.expects(:override).with({:http_pool => pool}).yields
       Puppet.expects(:override).with({:server => "myserver", :serverport => '123'}).yields
-      Puppet.expects(:override).with({:server => nil, :serverport => nil}).yields
       error = Net::HTTPError.new(400, 'dummy server communication error')
       Puppet::Node.indirection.expects(:find).raises(error)
-      @agent.expects(:run_internal).returns(nil)
-      @agent.run
+      expect{ @agent.run }.to raise_error(Puppet::Error)
```
Comment (joshcooper) [inline spec/unit/configurer_spec.rb]: Sorry just noticed this. I usually add (part of) the expected message, eg `/Could not select .... /`, so the test doesn't pass when it shouldn't.

Comment (melissa) [reply spec/unit/configurer_spec.rb]: oh derp, yes, thank you for catching this. I meant to go back and change that

Comment (melissa) [reply spec/unit/configurer_spec.rb]: Updated

Comment (joshcooper) [reply spec/unit/configurer_spec.rb]: Cool thanks!

**Comment by joshcooper** [comment]
Shoot, I just merged this and noticed the server list in the output isn't what I expected:

```
$ bundle exec puppet agent -td --server_list neptune,jupiter --server mars
...
Error: Could not run Puppet configuration client: Could not select a functional puppet master from server_list: '[["neptune", 8140], ["jupiter", 8140]]'
```

@jtappa when adding the `print` method for https://tickets.puppetlabs.com/browse/PUP-9495, could you update the `configurer` to call that method?

**Pull request merged**